### PR TITLE
Feature/us040/login system

### PR DESCRIPTION
**Pull request by mtslech**
### 🛠 Changes being made

Here give examples of the changes you've made in this pull request. Include an itemized list if you can. It'll help the reviewer

### ✨ Definition of Done (all the points need to be filled out before merging)

- [x] Meets Acceptance Criteria
- [ ] Tested
- [x] Code Review made by team 2 members
- [ ] UX approved
- [ ] Product Owner Approved

### 🧪 Non-Functional Requirements

- [ ] Automated tests should cover acceptance criterias (Unit tests, Integration/API/Component tests)
- [x] All input must be validated (client side & API)
- [x] Sensitive data should not be stored in the URL
- [ ] No secrets in codesource

### CODE REVIEW

File: product-web-app/tests/authService.test.ts [hunk 1]
```diff
@@ -1,16 +1,27 @@
-import { expect } from 'chai';
+import chai, { expect } from 'chai';
+import * as sinon from 'sinon';
+import sinonChai from 'sinon-chai';
```
Comment (mtslech) [inline product-web-app/tests/authService.test.ts]: To use them, you will need to install an additional package:
 npm install ts-sinon sinon-chai @types/sinon @types/sinon-chai --save-dev

**Comment by mtslech** [comment]
Login page:

image


After logging with valid credentials you should see token appear in cookies:

Screenshot 2023-10-06 at 15 38 43



Token can be decoded with this tool:
https://jwt.io/


Example:

Screenshot 2023-10-06 at 15 42 30

File: product-web-app/package.json [hunk 2]
```diff
@@ -27,25 +28,32 @@
...
     "eslint-plugin-import": "^2.27.5",
+    "jest": "^29.7.0",
     "mocha": "10.2.0",
     "prettier": "3.0.0",
     "sinon": "15.2.0",
+    "sinon-chai": "^3.7.0",
+    "ts-jest": "^29.1.1",
     "ts-mocha": "10.0.0",
     "ts-node": "10.9.1",
+    "ts-sinon": "^2.0.2",
```
Comment (ljercha) [inline product-web-app/package.json]: You are not using all the libraries you added (just sinon-chai)

File: product-web-app/service/authService.ts [hunk 3]
```diff
@@ -37,13 +38,19 @@ export default class AuthService {
...
-      const response = await axios.post('http://localhost:8080/api/login', user);
-
-      return response.data;
-    } catch (e) {
-      throw new Error('Could not login');
+      const apiResponse = await axios.post(API.LOGIN, user);
+      const token = apiResponse.data;
+
+      if (token) {
+        res.cookie('token', token, { httpOnly: true });
```
Comment (ljercha) [inline product-web-app/service/authService.ts:47]: Move setting a cookie to the controller, this method can just return extracted cookie from a body. You won't require those tests using sinon

File: product-web-app/service/authService.ts [hunk 4]
```diff
@@ -37,13 +38,19 @@ export default class AuthService {
     }
   }
 
-  async login(user: User): Promise<void> {
+  async login(user: User, res: Response): Promise<void> {
```
Comment (ljercha) [inline product-web-app/service/authService.ts]: Move setting a cookie to the controller, this method can just return extracted cookie from a body. You won't require those tests using sinon

File: product-web-app/service/authService.ts (hunk 3 above)
Comment (ljercha) [inline product-web-app/service/authService.ts]: set max-age to 1 hour
```suggestion
        res.cookie('token', token, { httpOnly: true });
```

File: product-web-app/controller/authController.ts [hunk 5]
```diff
@@ -3,6 +3,7 @@ import User from '../model/user.js';
 import AuthService from '../service/authService.js';
 import AuthValidator from '../service/authValidator.js';
 import logger from '../service/logger.js';
+import dayjs from 'dayjs';
```
Comment (mtslech) [inline product-web-app/controller/authController.ts:6]: You have to install Day.js library

npm install dayjs

File: product-backend/src/main/java/org/kainos/ea/service/AuthService.java [hunk 6]
```diff
@@ -48,7 +52,38 @@ public void register(User user) throws FailedToRegisterException, InvalidEmailEx
...
+        Algorithm algorithm = Algorithm.HMAC256("erggv45wv54c53xd345vcg4v54yv2");
+        Date expiry = DateUtils.addHours(new Date(), 1);
+
+        String token = JWT.create()
+                .withIssuer("auth0")
+                .withClaim("email", user.get().getEmail())
+                .withClaim("role", user.get().getRoleId())
+                .withExpiresAt(expiry)
+                .sign(algorithm);
+
```
Comment (Fluffyguy611) [inline product-backend/src/main/java/org/kainos/ea/service/AuthService.java:87]: Could split into own class/file

**Review by ljercha — APPROVED** [review]

File: product-web-app/views/login.html [hunk 7]
```diff
@@ -1,15 +1,5 @@
...
-
-<head>
-    <meta charset="utf-8" />
-    <title>{% block title %}{% endblock %}</title>
-    <meta name="viewport" content="width=device-width, initial-scale=1">
-    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-giJF6kkoqNQ00vy+HMDP7azOuL0xtbfIcaT9wjKHr8RbDVddVHyTfAAsrekwKmP1" crossorigin="anonymous">
-    <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/js/bootstrap.bundle.min.js" integrity="sha384-ygbV9kiqUc6oa4msXn9868pTtWMgiQaeYH7/t7LECLbyPA2x65Kgf80OJFdroafW" crossorigin="anonymous"></script>
-</head>
-<title>{% block title %}Login{% endblock %}</title>
-
```
Comment (Fluffyguy611) [inline product-web-app/views/login.html]: change back as the authorisation system is not yet ready and this was the correct way

**Review by Fluffyguy611 — APPROVED** [review]

**Pull request merged**